File: ArchitectureMVP/core/src/main/java/com/example/core/app/Configurator.java
package com.example.core.app;

import java.util.HashMap;

/**
 * File description.
 *
 * @author dsh
 * @date 2019-09-18
 */
public class Configurator {

    private static final HashMap<Object,Object> CONFIGS = new HashMap<>();

    private Configurator(){
        CONFIGS.put(ConfigKeys.API_HOST,false);
    }

    private static class ConfiguratorHolder{
        private static Configurator INSTANCE = new Configurator();
    }

    public static Configurator getInstance(){
        return ConfiguratorHolder.INSTANCE;
    }

    final HashMap<Object,Object> getConfigs(){
        return CONFIGS;
    }

    public final Configurator withAPIHost(String apiHost){
        CONFIGS.put(ConfigKeys.API_HOST,apiHost);
        return this;
    }

    //配置完成
    public final void configure(){
        CONFIGS.put(ConfigKeys.CONFIG_READY.name(),true);
    }

    //检查配置是否完成
    private void checkConfiguration(){
        final boolean isReady=(boolean)CONFIGS.get(ConfigKeys.CONFIG_READY.name());
        if(!isReady){
            throw new RuntimeException("Configuration is not ready,call configure()");
        }
    }

    final<T> T getConfiguration(Object key){
        checkConfiguration();
        final Object value=CONFIGS.get(key);
        if(value==null){
            throw new NullPointerException(key.toString()+"IS NULL");
        }
        return (T)value;
    }

}

File: MangoKotlin/core/src/main/java/com/example/core/Configurator.kt
package com.example.core

import android.os.Handler
import java.util.*

/**
 * File description.
 *
 * @author dsh
 * @date 2019-08-21
 */

class Configurator {
    companion object {
        private var LEMON_CONFIGS = HashMap<Any, Any>()
        private var HANDLER = Handler()
    }

    constructor() {
        LEMON_CONFIGS[ConfigKeys.CONFIG_READY] = false
        LEMON_CONFIGS[ConfigKeys.HANDLER] = HANDLER
    }

    fun withApiHost(host: String): Configurator {
        LEMON_CONFIGS[ConfigKeys.API_HOST] = host
        return this
    }

    internal fun <T> getConfiguration(key: Any): T {
        checkConfiguration()
        val value = LEMON_CONFIGS[key] ?: throw NullPointerException(key.toString() + "is NULL")
        return value as T
    }

    private fun checkConfiguration() {
        val isReday = LEMON_CONFIGS[ConfigKeys.CONFIG_READY] as Boolean
        if (!isReday) {
            throw RuntimeException("Configuration is no ready,please call configure!")
        }
    }

}

